### ISSUE TITLE
[Feature Req] Add image sharpen and custom color filters

### ISSUE THREAD

**Issue by Ero-gamer** [KotatsuApp/Kotatsu#962]
### Describe your suggested feature

Hi! If possible add image sharpen and custom color filters (like the one Tachiyomi uses with overlays..etc) since image contrast filter use cases are limited and few.

### Acknowledgements

- [X] This is not a duplicate of an existing issue. Please look through the list of [open issues](https://github.com/KotatsuApp/Kotatsu/issues) before creating a new one.

**Comment by galpt** [comment]
I believe image sharpening is heavier compared to simple overlays, it will definitely consume the battery real quick.
You're basically telling your phone to edit every manga image to look sharper.

It'll be easier to implement the sharpening from the img optimization proxy feature.
The compute can be offloaded to the proxy and Kotatsu only need to display images.
But I haven't seen proxies that can apply color overlays, only compression & sharpening, so overlays might still need to be done by Kotatsu.

**Comment by galpt** [comment]
I believe sharpening and coloring can be implemented 100% from the 0ms.dev proxy.
@Ero-gamer could you give a link or two about what you're expecting from this?

Sharpening is straight to the point, but color filters - I might need to see what color filters you're referring to and how they look.
Since I'm one of the devs who maintain the 0ms.dev img proxy, I can give sharpening a try.
I'll let you know here if it's a success.

**Comment by galpt** [comment]
@Ero-gamer 
So this is a quick example after implementing sharpening to via the proxy.
Before:
https://cdn.myanimelist.net/images/anime/1932/142249l.jpg
After:
https://x.0ms.dev/q70/https://cdn.myanimelist.net/images/anime/1932/142249l.jpg?sharp=true

Well, I'd say manga images are already optimized, so adding sharpening effect will make them look worse.
You can experiment with more images you could find on the Internet and decide if img sharpening is worth adding to Kotatsu.

**Comment by Ero-gamer** [comment]
Here is an example of manhwa image with and without filter: 

![Screenshot_2024-07-10-00-35-35-42](https://github.com/KotatsuApp/Kotatsu/assets/146583787/ae954a7f-95ad-4414-ad2f-f8280cee7ad1)
![Screenshot_2024-07-10-00-36-10-87](https://github.com/KotatsuApp/Kotatsu/assets/146583787/ee945f13-f201-4c92-a0b3-1f0b3c5f8a2a)

As you can see, original image is overbright which mess up colors, I have read tons of webtoons/manhwa and most of them suffer from overbright colors, some of them are affected slightly and the rest affected greatly.
This feature is available on original Tachiyomi and Mihon as overlay color filter

![Screenshot_2024-07-10-00-36-26-69](https://github.com/KotatsuApp/Kotatsu/assets/146583787/897cd7d4-3cb4-44a5-801b-0fc133f61b83)

**Comment by Ero-gamer** [comment]
> @Ero-gamer So this is a quick example after implementing sharpening to the proxy. Before: https://cdn.myanimelist.net/images/anime/1932/142249l.jpg After: https://x.0ms.dev/q70/https://cdn.myanimelist.net/images/anime/1932/142249l.jpg?sharp=true
> 
> Sharpening will only be used if `?sharp=true` is added at the end of the URL. By default proxy will just forward images as-is.
> 
> Well, I'd say manga images are already optimized, so adding sharpening effect will make them look worse. You can experiment with more images you could find on the Internet and decide if img sharpening is worth adding to Kotatsu.

I compared both images and they look identical, no sharpen filter. Sharpen filter is great on certain blurry and low res images especially if it's a high quality sharpen filter like lanczos3 or lanczos4 sharpen filter.

**Comment by galpt** [comment]
That overlay feature's indeed can be useful for some manga if the devs implement it.

> > @Ero-gamer So this is a quick example after implementing sharpening to the proxy. Before: https://cdn.myanimelist.net/images/anime/1932/142249l.jpg After: https://x.0ms.dev/q70/https://cdn.myanimelist.net/images/anime/1932/142249l.jpg?sharp=true
> > Sharpening will only be used if `?sharp=true` is added at the end of the URL. By default proxy will just forward images as-is.
> > Well, I'd say manga images are already optimized, so adding sharpening effect will make them look worse. You can experiment with more images you could find on the Internet and decide if img sharpening is worth adding to Kotatsu.
> 
> I compared both images and they look identical, no sharpen filter. Sharpen filter is great on certain blurry and low res images especially if it's a high quality sharpen filter like lanczos3 or lanczos4 sharpen filter.

About the sharpening, it's hard because people's screens and eyes are different, so person A might see it as something better while person B might not notice a different or even see it as worse than the original image.

This is another example, should be noticable.
The sharpening effect shouldn't be much because like I mentioned, it'll make the image looks even stranger because sharpening is playing with brightness & contrast of the images.

Original: https://cdn.myanimelist.net/images/anime/1567/138687l.jpg
After sharpening: https://x.0ms.dev/q70/https://cdn.myanimelist.net/images/anime/1567/138687l.jpg?sharp=true

There's another term that's called "upscaling" where you'll need a GPU to make every image looks clearer and better than the original, but that'll cost so much just for reading manga, and nobody wants to spend so much money just to read manga everyday.

It'll be expensive to do upscaling via proxy, so this should be done on people's phones with their own mobile GPUs.

The original before upscaling:
![138687l](https://github.com/KotatsuApp/Kotatsu/assets/34034590/67acbb27-fccd-4660-bcfe-b22f1830cd35)

The image after upscaled:
![138687l](https://github.com/KotatsuApp/Kotatsu/assets/34034590/41b4d55b-d792-4912-8c62-d103cafc2e6b)

**Comment by HironoKenta** [comment]
@Ero-gamer IDK, maybe I'm wrong, but lanczos isn't a sharp filter but a scaler like bicubic, spline, jinc.

Maybe it can be implemented as an optional sharpen scaler in place of the default one?

**Comment by Ero-gamer** [comment]
> @Ero-gamer IDK, maybe I'm wrong, but lanczos isn't a sharp filter but a scaler like bicubic, spline, jinc.
> 
> Maybe it can be implemented as an optional sharpen scaler in place of the default one?

You are correct, it's image scaling algorithm with sharp filter, it can be implemented as image scaling and subsampling algorithm.

**Comment by HironoKenta** [comment]
@Ero-gamer but if it don't change image resolution it usually just do nothing.
A simple sharpen filter can be implemented as a 0,-1,0,-1,4,-1,0,-1,0 or -1,-2,-1,-2,12,-2,-1,-2,-1 convolution matrix.

**Comment by radry** [comment]
+1 for color filter. It should be easy to implement as it's just an overlay. There even is a menu "color correction" in  the app but there is only brightness and contrast control, no "color" correction yet.
